main.cpp

```diff
@@ -59,26 +74,33 @@ static void matrix_multiply(Matrix &out, Matrix const &lhs, Matrix const &rhs) {
         std::cerr << "matrix_multiply: shape mismatch" << std::endl;
         throw;
     }
-    out.reshape(nx, ny);
+    out.reshape({nx, ny}, 0);
 
     // 这个循环为什么不够高效？如何优化？ 15 分
-#pragma omp parallel for collapse(2)
-    for (int y = 0; y < ny; y++) {
-        for (int x = 0; x < nx; x++) {
-            out(x, y) = 0;  // 有没有必要手动初始化？ 5 分
-            for (int t = 0; t < nt; t++) {
-                out(x, y) += lhs(x, t) * rhs(t, y);
+
+    constexpr size_t grain = 8;
+    tbb::parallel_for(tbb::blocked_range3d<size_t>((size_t)0, ny, grain,(size_t) 0, nt, grain,(size_t) 0, nx, grain),
+            [&](tbb::blocked_range3d<size_t> r){
+        for(int y = r.pages().begin();  y < r.pages().end(); ++y)
+        {
+            for(int x = r.cols().begin();  x < r.cols().end(); ++x)
+            {
+                for(int t = r.rows().begin(); t < r.rows().end(); ++t)
+                {
+                    out(x, y) += lhs(x, t) * rhs(t, y);
+                }
             }
         }
-    }
+    }, tbb::auto_partitioner{});
     TOCK(matrix_multiply);
 }
 
 // 求出 R^T A R
 static void matrix_RtAR(Matrix &RtAR, Matrix const &R, Matrix const &A) {
     TICK(matrix_RtAR);
     // 这两个是临时变量，有什么可以优化的？ 5 分
-    Matrix Rt, RtA;
+
+    Matrix Rt(R.shape()), RtA(A.shape());
```

Comment:
```suggestion
    static thread_local Matrix Rt, RtA;
```
上课听了吗？这样才能在多次调用中重用内存。